### PR TITLE
Collision Prevention: support multiple sensors and frames

### PR DESCRIPTION
**Pull request by baumanta**
**Describe problem solved by the proposed pull request**
This PR enables the use of multiple independent obstacle distance messages. It generalizes the the way the collision prevention library gathers data, such that an arbitrary number and combination of `obstacle_distance` and `distance_sensor` messages can be used. It also allows for `obstacle_distance` messages to arrive in either world or body frame, whereas before only world frame messages were supported.
This is done by collecting all incoming sensor messages in an internal obstacle map in body frame, which is then used to calculated the constrained setpoint.
This PR also automatically limits the maximum speed of the vehicle to be able to brake at the largest sensor range.

**Test data / coverage**
This was SITL tested together with the avoidance SITL package and there have been added a bunch of unit tests. Flight tests pending.

### CODE REVIEW

**Comment by baumanta** [comment]
Test flight with the avoidance system providing an obstacle_distance message. Flight behavior looked the same, no regression.
https://logs.px4.io/plot_app?log=6a5b239a-b1c3-47a2-aebe-c21dfaeb3506

File: src/lib/CollisionPrevention/CollisionPrevention.cpp [hunk 1]
```diff
@@ -58,171 +72,212 @@ CollisionPrevention::~CollisionPrevention()
...
 
-					if (wrap_bin < 0) {
-						// wrap bin index around the array
-						wrap_bin = (int)floor(360.f / obstacle.increment) + bin;
-					}
+		// rotate vehicle attitude into the sensor body frame
+		const int map_bins_used = 360.f / _obstacle_map_body_frame.increment;
+		matrix::Quatf attitude_sensor_frame = vehicle_attitude;
+		attitude_sensor_frame.rotate(Vector3f(0.f, 0.f, sensor_yaw_body_rad));
+		float attitude_sensor_frame_pitch = cosf(Eulerf(attitude_sensor_frame).theta());
```
Comment (jkflying) [inline src/lib/CollisionPrevention/CollisionPrevention.cpp]:
```suggestion
		float sensor_dist_scale = cosf(Eulerf(attitude_sensor_frame).theta());
```

File: msg/obstacle_distance.msg [hunk 2]
```diff
@@ -1,6 +1,10 @@
 # Obstacle distances in front of the sensor.
 uint64 timestamp		# time since system start (microseconds)
 
+uint8 frame		#Coordinate frame of reference for the yaw rotation and offset of the sensor data. Defaults to MAV_FRAME_GLOBAL, which is North aligned. For body-mounted sensors use MAV_FRAME_BODY_FRD, which is vehicle front aligned.
+uint8 MAV_FRAME_GLOBAL = 0
```
Comment (mhkabir) [inline msg/obstacle_distance.msg]: This should be MAV_FRAME_LOCAL_NED.

File: src/lib/CollisionPrevention/CollisionPrevention.cpp [hunk 3]
```diff
@@ -58,171 +72,212 @@ CollisionPrevention::~CollisionPrevention()
 	}
 }
 
-void CollisionPrevention::_updateOffboardObstacleDistance(obstacle_distance_s &obstacle)
+hrt_abstime CollisionPrevention::getTime()
```
Comment (mhkabir) [inline src/lib/CollisionPrevention/CollisionPrevention.cpp]: Why is this needed?

Comment (jkflying) [reply src/lib/CollisionPrevention/CollisionPrevention.cpp]: So that we can fake the time in the tests to check that timeouts work, and that old data is dropped.

File: msg/obstacle_distance.msg (hunk 2 above)
Comment (jkflying) [reply msg/obstacle_distance.msg]: See https://github.com/mavlink/mavlink/pull/1214

Agreed we should support LOCAL_NED too, although in this case it will be exactly the same as GLOBAL because we just care about the yaw offset.

**Comment by dagar** [comment]
> CPU usage is really high, here are some quick fixes for that.

Can you be a little lazier with some of the work?

For example, copying every single distance_sensor message every iteration regardless of update.

https://github.com/PX4/Firmware/blob/5022162392916023ad2d1d07633a7423c1d1092f/src/lib/CollisionPrevention/CollisionPrevention.cpp#L150-L153

The obstacle_distance message isn't force updated every iteration, but the data is still added again.

https://github.com/PX4/Firmware/blob/5022162392916023ad2d1d07633a7423c1d1092f/src/lib/CollisionPrevention/CollisionPrevention.cpp#L173-L183

**Comment by jkflying** [comment]
@dagar we're running out of flash on stackcheck FMUV5. Should we try to do something here, or is there somewhere else we can fix it?

**Comment by dagar** [comment]
Should be fixed in master.

**Comment by jkflying** [comment]
V2 flash...

**Comment by MaEtUgR** [comment]
@baumanta @jkflying I checked what uses up all the flash and found that replacing the `fmod` call saves all the necessary flash which must be more than 900 bytes.

So in minimal you can just apply my last commit from https://github.com/PX4/Firmware/pull/new/sensor-combination here to use the existing wrap implementation of the matrix library or make it nicer.

I planned to do a general wrap function for any upper and lower limit but did the same mistake with fmod because it looks so nice. Here's the draft that has yet to be finished: https://github.com/PX4/Matrix/pull/71

**Comment by LorenzMeier** [comment]
`fmod` should NOT be used for angles in an embedded system. We know that the angles are very few iterations off (1-2) if the system is in any sane condition. If you get 239393494 degrees as input you might as well just throw NaN, because you know that the input was invalid. Fast math code does not use boilerplate math functions. The same is true for `pow()` - you should NEVER EVER use that. Anyone using that for linear algebra doesn't understand anything about programming. And I'm saying that so pointed to make sure anyone finding this comment spends the time to read up on this and ask their co-workers.

**Comment by MaEtUgR** [comment]
@LorenzMeier I had to learn it the hard way. By now I can say you're right with confidence. I'll finish my matrix pr **without** `fmod` now such that it can be used directly with degrees here.

**Comment by jkflying** [comment]
This was me, I was under the impression that there would be an embedded-friendly version in the toolchain already, otherwise it wouldn't be available.

Let's make either a safe implementation of fmod/pow/log/whatever, or at least a static assert that fails at compile time if it is used. Otherwise we have to repeat this every time there is a new developer, hope we catch it in review, etc. I believe we had a very similar discussion last week in the dev call over usage of std::vector slipping in, and how to prevent it.

**Comment by julianoes** [comment]
@LorenzMeier 
> Anyone using that for linear algebra doesn't understand anything about programming.

I don't like this tone, let's be respectful as is in our [code of conduct](https://github.com/PX4/Firmware/blob/master/CODE_OF_CONDUCT.md#our-standards).

**Comment by jkflying** [comment]
This is overflowing V2 flash again, both from the Matrix update and from merging master.

**Comment by jkflying** [comment]
Ok this is working again, we just need the matrix library PR to be merged.
